### PR TITLE
LYN-7693 Rename and move vegetation reference shape.

### PR DESCRIPTION
**Pull request by sphrose**
Signed-off-by: sphrose <82213493+sphrose@users.noreply.github.com>

### CODE REVIEW

File: Gems/LmbrCentral/Code/Source/Shape/EditorReferenceShapeComponent.h [hunk 1]
```diff
@@ -8,24 +8,24 @@
...
+        using BaseClassType = EditorWrappedComponentBase<ReferenceShapeComponent, ReferenceShapeConfig>;
         AZ_EDITOR_COMPONENT(EditorReferenceShapeComponent, EditorReferenceShapeComponentTypeId, BaseClassType);
         static void Reflect(AZ::ReflectContext* context);
 
-        static constexpr const char* const s_categoryName = "Vegetation";
-        static constexpr const char* const s_componentName = "Vegetation Reference Shape";
+        static constexpr const char* const s_categoryName = "Shape";
+        static constexpr const char* const s_componentName = "Reference Shape";
         static constexpr const char* const s_componentDescription = "Enables the entity to reference and reuse shape entities";
         static constexpr const char* const s_icon = "Editor/Icons/Components/Vegetation.svg";
```
Comment (mbalfour-amzn) [inline Gems/LmbrCentral/Code/Source/Shape/EditorReferenceShapeComponent.h:27]: This icon should also get replaced.

**Review by mbalfour-amzn — COMMENTED** [review]
Just to verify, are there any python files that reference "Vegetation Reference Shape" by name?

**Comment by sphrose** [comment]
> Just to verify, are there any python files that reference "Vegetation Reference Shape" by name?

No

File: Gems/LmbrCentral/Code/Source/Shape/EditorReferenceShapeComponent.h (hunk 1 above)
Comment (sphrose) [reply Gems/LmbrCentral/Code/Source/Shape/EditorReferenceShapeComponent.h:27]: Is there one available?

Comment (mbalfour-amzn) [reply Gems/LmbrCentral/Code/Source/Shape/EditorReferenceShapeComponent.h:27]: I guess just use Editor/Icons/Components/Component_Placeholder.svg to match what you're using on the next line for the viewport icon.  Longer term we'll need to request an icon for this component.

File: Gems/LmbrCentral/Code/Source/Shape/EditorReferenceShapeComponent.h [hunk 2]
```diff
@@ -24,7 +24,7 @@ namespace LmbrCentral
         static constexpr const char* const s_categoryName = "Shape";
         static constexpr const char* const s_componentName = "Reference Shape";
         static constexpr const char* const s_componentDescription = "Enables the entity to reference and reuse shape entities";
-        static constexpr const char* const s_icon = "Editor/Icons/Components/Vegetation.svg";
+        static constexpr const char* const s_icon = "Icons/Components/Viewport/Component_Placeholder.svg";
```
Comment (mbalfour-amzn) [inline Gems/LmbrCentral/Code/Source/Shape/EditorReferenceShapeComponent.h:27]: This shouldn't have "Viewport" in the path, it uses a different icon than the Viewport icon.  The regular one is white squares on a transparent background, the viewport one is white squares on a grey background.
image
image

**Review by invertednormal — APPROVED** [review]

File: Gems/LmbrCentral/Code/Source/Shape/EditorReferenceShapeComponent.h [hunk 3]
```diff
@@ -8,24 +8,24 @@
...
     public:
-        using BaseClassType = EditorVegetationComponentBase<ReferenceShapeComponent, ReferenceShapeConfig>;
+        using BaseClassType = EditorWrappedComponentBase<ReferenceShapeComponent, ReferenceShapeConfig>;
         AZ_EDITOR_COMPONENT(EditorReferenceShapeComponent, EditorReferenceShapeComponentTypeId, BaseClassType);
         static void Reflect(AZ::ReflectContext* context);
 
-        static constexpr const char* const s_categoryName = "Vegetation";
-        static constexpr const char* const s_componentName = "Vegetation Reference Shape";
+        static constexpr const char* const s_categoryName = "Shape";
+        static constexpr const char* const s_componentName = "Reference Shape";
```
Comment (mbalfour-amzn) [inline Gems/LmbrCentral/Code/Source/Shape/EditorReferenceShapeComponent.h:25]: I just talked with UX, and they suggested the display name should be "Shape Reference" instead of "Reference Shape", which I'd agree makes more sense.  We can leave the file names the same, but just change the display name here.

File: Gems/LmbrCentral/Code/Source/Shape/EditorReferenceShapeComponent.h (hunk 2 above)
Comment (sphrose) [reply Gems/LmbrCentral/Code/Source/Shape/EditorReferenceShapeComponent.h]: Changed

File: Gems/LmbrCentral/Code/Source/Shape/EditorReferenceShapeComponent.h (hunk 3 above)
Comment (sphrose) [reply Gems/LmbrCentral/Code/Source/Shape/EditorReferenceShapeComponent.h]: Changed

File: Gems/LandscapeCanvas/Code/Source/Editor/MainWindow.cpp [hunk 4]
```diff
@@ -226,6 +227,22 @@ namespace LandscapeCanvasEditor
...
+            {
+                const auto& componentPair = categoryIt->second.find(preferredComponentPair.second);
+
+                if (componentPair != categoryIt->second.end())
+                {
+                    return componentPair->second->m_typeId;
+                }
+            }
+        }
+
```
Comment (mbalfour-amzn) [inline Gems/LandscapeCanvas/Code/Source/Editor/MainWindow.cpp:245]: This might _work_ for the one example we have, but I don't think it's correct.  I believe the way this method works is that it has an input set of categories and component names that have already been filtered down to be all the potential matches for the component service we're trying to match.  So, this method is trying to pick the best component out of that set.
What I think we _want_ is to say "If we have a specific preferred component that exists in the set, return that.  If not, if we have a specific preferred category, pick the first component from that category.  If not, just return the first component that matches."
What I think we're saying with the above is "If we have a specific preferred category, look in that category for any of our preferred components.  If we don't, look in the first category for any of our preferred components.  Otherwise, return the first component in that category."
Basically, I think the preferredComponentByCategory loop should appear before the preferredCategory loop.  Also, since it's checking everything in the preferredComponentByCategory, that could just be a static const array instead of an unordered_map.  We're never performing lookups in it, so an array would be a better choice than a map.

File: Gems/LmbrCentral/Code/Source/Shape/ReferenceShapeComponent.cpp [hunk 5]
```diff
@@ -27,7 +27,7 @@ namespace Vegetation
             if (edit)
             {
                 edit->Class<ReferenceShapeConfig>(
-                    "Vegetation Reference Shape", "")
+                    "Reference Shape", "")
```
Comment (mbalfour-amzn) [inline Gems/LmbrCentral/Code/Source/Shape/ReferenceShapeComponent.cpp:30]: I think this should be "Shape Reference"?

**Review by mbalfour-amzn — CHANGES_REQUESTED** [review]

File: Gems/LmbrCentral/Code/Source/Shape/EditorReferenceShapeComponent.h [hunk 6]
```diff
@@ -8,24 +8,24 @@
 
 #pragma once
 
-#include <Components/ReferenceShapeComponent.h>
-#include <Vegetation/Editor/EditorVegetationComponentBase.h>
+#include<LmbrCentral/Component/EditorWrappedComponentBase.h>
```
Comment (moraaar) [inline Gems/LmbrCentral/Code/Source/Shape/EditorReferenceShapeComponent.h:11]: nit: missing space beween `include` and  `<`

File: Gems/LmbrCentral/Code/Mocks/LmbrCentral/Shape/MockShapes.h [hunk 7]
```diff
@@ -56,5 +56,73 @@ namespace UnitTest
         MOCK_METHOD1(GenerateRandomPointInside, AZ::Vector3(AZ::RandomDistributionType randomDistribution));
         MOCK_METHOD3(IntersectRay, bool(const AZ::Vector3& src, const AZ::Vector3& dir, float& distance));
     };
+
+    class MockShape : public LmbrCentral::ShapeComponentRequestsBus::Handler
+    {
+    public:
+        AZ::Entity m_entity;
+        mutable int m_count = 0;
```
Comment (moraaar) [inline Gems/LmbrCentral/Code/Mocks/LmbrCentral/Shape/MockShapes.h:64]: it doesn't seem necessary for `m_count` to be mutable, I haven't seen it modified in a `const` environment.

**Review by moraaar — APPROVED** [review]
LGTM. Just a couple of minor comments. I also agree with @mbalfour-amzn comments, approving considering his feedback will be covered. Cheers.

File: Gems/LandscapeCanvas/Code/Source/Editor/MainWindow.cpp [hunk 8]
```diff
@@ -201,13 +203,12 @@ namespace LandscapeCanvasEditor
     {
         using namespace AzToolsFramework;
 
-        static const QStringList preferredCategories = {
-            "Vegetation",
-            "Atom"
-        };
+        static const QStringList preferredCategories = { "Vegetation", "Atom" };
+
+        static const AZStd::unordered_map<AZStd::string, QString> preferredComponentByCategory = { { "Shape", "Shape Reference" } };
```
[the remark this answers is not in the document]
Comment (sphrose) [reply Gems/LandscapeCanvas/Code/Source/Editor/MainWindow.cpp:208]: Ooops, changed.

File: Gems/LandscapeCanvas/Code/Source/Editor/MainWindow.cpp (hunk 4 above)
Comment (sphrose) [reply Gems/LandscapeCanvas/Code/Source/Editor/MainWindow.cpp:245]: I can think of arguments for both ways round, but switching the order probably makes most sense. I've left it as a unordered_map, but put a lookup in instead, which is what I intended in the first place.

File: Gems/LmbrCentral/Code/Source/Shape/ReferenceShapeComponent.cpp (hunk 5 above)
Comment (sphrose) [reply Gems/LmbrCentral/Code/Source/Shape/ReferenceShapeComponent.cpp:30]: Fixed

File: Gems/LmbrCentral/Code/Source/Shape/EditorReferenceShapeComponent.h (hunk 6 above)
Comment (sphrose) [reply Gems/LmbrCentral/Code/Source/Shape/EditorReferenceShapeComponent.h:11]: Fixed

File: Gems/LmbrCentral/Code/Mocks/LmbrCentral/Shape/MockShapes.h (hunk 7 above)
Comment (sphrose) [reply Gems/LmbrCentral/Code/Mocks/LmbrCentral/Shape/MockShapes.h:64]: This is just code moved from vegetation_mocks, although I forgot to delete the old version (done now)/ I've removed mutable as well.

File: Gems/LandscapeCanvas/Code/Source/Editor/MainWindow.cpp [hunk 9]
```diff
@@ -201,15 +203,29 @@ namespace LandscapeCanvasEditor
...
+
+        if (preferredComponentPair != preferredComponentByCategory.end())
+        {
+            const auto& componentPair = componentDataTable.begin()->second.find(preferredComponentPair->second);
+
+            if (componentPair != componentDataTable.begin()->second.end())
+            {
+                return componentPair->second->m_typeId;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (sphrose) [reply Gems/LandscapeCanvas/Code/Source/Editor/MainWindow.cpp]: I see what you mean, yes that would happen. I've redone it, so that it covers that case and anything else I can think of.

**Review by mbalfour-amzn — APPROVED** [review]
